move `runJob` to its own package

mmjconolly opened the pull request:
Wake has long provided a default mechanism for choosing a `Runner` for a `Job`.
It uses a scoring function to pick the runner, and allows multiple runners to modify the plan before it actually runs.
It can become difficult to reason about which runner will modify or run any particular job, so alternative implementations can become desired.

This PR moves `runJob` and `runIncrementalJob` into a package called `plan_scorer`.
This package could eventually move into the wake-extras repository, perhaps with other schemes for choosing a Runner.

Wake jobs that are used in wake's own build are transitioned to using `runJobWith`.
Documentation has some minimal modifications.

End users can opt-in to continuing to use `runJob`/`runIncrementalJob` by adding:
```
from plan_scorer import _
```
to each file that needs it. This allows them to specify an alternative `runJob` implementation in other files.

As an immediate migration, a project could add the following to their source code:
```
global def runJob (p: Plan) : Job =
    from plan_scorer import runJob
    runJob p

 ## and then adding `from plan_scorer import topic runner` anywhere a `Runner` is `publish`ed.
```

Alternatively, a project with very simple needs could add a shorthand such as `global def run = runJobWith defaultRunner` with the caveat that `LocalOnly` in the `Plan` would be ignored.

JakeSiFive commented on share/wake/lib/system/job.wake:
```diff
@@ -665,7 +665,7 @@ export topic runner: Runner
     localRunner, defaultRunner, Nil
 
-def runJobImp label cmd env dir stdin res usage finputs foutputs vis pers run (LogLevel echo) (LogLevel stdout) (LogLevel stderr) isatty =
+export def runJobImp label cmd env dir stdin res usage finputs foutputs vis pers run (LogLevel echo) (LogLevel stdout) (LogLevel stderr) isatty =
```
This is a bit more concerning to export. Is there a reason that you can't use `runJobWith localRunner` for execution needs? This exposes a bit more the internal functionality than I'm currently comfortable with.

mmjconolly replied:
Yeah I can try to make it use `runJobWith`, though perhaps with the runner it chooses from the scoring

JakeSiFive replied:
I'm not sure I understand. I just meant I don't want to export runJobImp just yet (or if we export it we should think about it more). It seems to me that any external runner can be implemented in terms of `runJobWith localRunner`

mmjconolly replied:
I think its easier for now if we export runJobImp, as it operates on the level of RunnerInput/Output rather than Plan (like runJobWith)

JakeSiFive replied:
Ok, that seems fine for now. I'll try to think of a better wrapper at some point.

JakeSiFive commented on share/wake/lib/versions/v34.wake:
```diff
@@ -0,0 +1,10 @@
+package v34_wake
```
Instead of making a v34 lib, its just as easy to switch packages from v34 to importing plan_scorer so I don't think this v34 buys us anything.

JakeSiFive reviewed (commented):
95% happy with this!

ag-eitilt commented on share/wake/lib/system/plan_scorer.wake:
```diff
@@ -0,0 +1,74 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+package plan_scorer
```
Is there anything left in the main package which uses the `topic runner`?  If not we should move that definition here as well.  Otherwise, calling this `plan_scorer` seems a bit internal-details-y.

mmjconolly replied:
Ah I forgot about the topic

JakeSiFive replied:
I had just blindly assumed it had moved over! Thanks Sam!

mmjconolly commented on share/wake/lib/system/job.wake:
```diff
@@ -706,59 +706,6 @@ def runJobImp label cmd env dir stdin res usage finputs foutputs vis pers run (L
 export def runJobWith (Runner _ _ run) (Plan label cmd vis env dir stdin stdout stderr echo pers _ res _ usage finputs foutputs isatty) =
     runJobImp label cmd env dir stdin res usage finputs foutputs vis pers run echo stdout stderr isatty
```
>   It would be a more involved migration, but do we want to take this opportunity to remove that marked-as-deprecated field?

The `score` functions look at `LocalOnly`, which deprecated its still in heavy use.
I'd like to keep this PR focused on removing the current `runJob` from the default package.

ag-eitilt replied:
Definitely, since they're still being used (those functions were the one place I forgot about checking).  The package split makes me all the more eager to remove the `LocalOnly` since it's not going to be used within `package wake` and having it as a core part of the tuple for an opt-in setup seems a bit ugly.  Definitely not part of this PR, though, and I'm perfectly part of it not being part of this same release.

When we do remove it, though, I wonder if we want to refactor the type to be `Plan a` so that project-specific infrastructure can attach and manipulate specific metadata.  The `package plan_scorer` would use `Plan Boolean` for local-only jobs (or the more generic `Plan ScoreInput`), for example.

mmjconolly commented:
Changes since last time are in the added commits
* `runIncrementalJob` is removed
* `runner` topic is moved into plan_scorer package
* and a rebase against master branch

JakeSiFive reviewed (approved):
I'm so excited to get rid of runJobIncremental (thanks @jackkoenig)

davidmlw commented:
> 

Why `runIncrementalJob` has no users? I use it for latex jobs, which compile my resume:) 
What is the elegant replace for runIncrementalJob?

JakeSiFive commented:
You can go back into an older version and copy it into your local wake code. We didn't remove to core features that made that work we just dropped support for it.

ag-eitilt commented:
This could just be me not understanding the TeX flow as it interacts with Wake, but what do you gain from using `runIncrementalJob`?  Given how complex the TEXMF structure is, I agree that trying to compile your resume in a sandbox isn't viable, but I would expect `runJobWith localRunner` to work for you.

davidmlw commented:
> 

Latex may need run the second time to use last output as input and update the same output files. This sceneria makes runIncrementalJob meaningful. Yes, localRunner works but cannot track the input/output dependencies. localRunner and fuseRunner will complain file hash errors or only run the first time and refuse to run the second time.

ag-eitilt commented:
Ah, that would be difficult with Wake's model; I've only ever used ConTeXt which handles all the rerun-if-necessary itself so you only ever have to call it once, and was under the impression LaTeX was also smart about doing so.  In that case, I do think Jake's solution of "download the [last version of the file](https://github.com/sifive/wake/blob/8399d249a61bc5e14a520e819489927a2a0a94e3/share/wake/lib/system/incremental.wake) into your workspace" is the best way to fix things; it should just be able to be dropped in alongside and work with no changes to the rest of your code, but I've not verified that myself.
